### PR TITLE
[JBPM-8305] getProcessInstance(readOnly=true) in workflow paused on task activity in subprocess causes extra run of first WorkItem after subprocess node.

### PR DESCRIPTION
**Pull request by elguardian**
disconnect again the process instance when it is in readOnly mode

### CODE REVIEW

**Comment by elguardian** [comment]
this ok to test.

**Comment by fe-c** [comment]
Why [JBPM-8363] and not [JBPM-8305] as issue id?

**Comment by elguardian** [comment]
@fe-c my bad.

**Comment by elguardian** [comment]
this ok to test.

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java [hunk 1]
```diff
@@ -92,8 +92,8 @@ public void testAutoCompleteCheckActivationData() {
                                 @Override
                                 public Boolean execute(Context context) {
                                     KieSession ksession = ((RegistryContext) context).lookup( KieSession.class );
-                                    
-                                    WorkflowProcessInstance pi = (WorkflowProcessInstance) ksession.getProcessInstance(processInstanceId, true);
+
```
Comment (mswiderski) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: Why is this change needed?

Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: getVariable of node instance requires read/write process. if you get this read only the test will fail.

Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: why getVariable requires read/write process?

Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: @mswiderski 
I cannot tell you why some operations were designed to work with read only or read / write. That is not documented anywhere.
The reason is when you disconnect the process this condition is fullyfilled
https://github.com/kiegroup/jbpm/blob/7.20.0.Final/jbpm-flow/src/main/java/org/jbpm/process/instance/impl/ProcessInstanceImpl.java#L97

causing failure in here
https://github.com/kiegroup/jbpm/pull/1471/files/e13dc20a8a8be0513b9aab4a1553d52703b31b3f#diff-6c5262fecaf86cc99369126786aad13cR97

Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: what I am asking is why this is needed? It wasn't needed before and I don't see a point in writing anything to process instance info when just reading it's variables. So could you point me what is actually making to fail?

Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: the problem in the jira is that the listener were registed for a read only process instance. This was because while in some parts of the code the process instance reconnected then after being used was disconnected again.

That was happening here
https://github.com/kiegroup/jbpm/blob/7.19.0.Final/jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java#L174

The collateral effect is that some operations that were being relying on the process instance were connected now are failing. Disconnecting the process instance causes the state variables to be null therefore you cannot navigate to the getProcess (this happens internally during getVariable)

stack trace for the getting variable
~~~
at org.jbpm.process.instance.impl.ProcessInstanceImpl.getProcess(ProcessInstanceImpl.java:97)
	at org.jbpm.workflow.instance.impl.WorkflowProcessInstanceImpl.getWorkflowProcess(WorkflowProcessInstanceImpl.java:292)
	at org.jbpm.workflow.instance.impl.WorkflowProcessInstanceImpl.getNodeContainer(WorkflowProcessInstanceImpl.java:124)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.getNode(NodeInstanceImpl.java:134)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.resolveContext(NodeInstanceImpl.java:434)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.resolveContextInstance(NodeInstanceImpl.java:441)
	at org.jbpm.workflow.instance.impl.NodeInstanceImpl.getVariable(NodeInstanceImpl.java:487)
~~~

Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: did you track down what is calling disconnect during the execution? It should only be invoked upon completion ... usually transaction completion.

Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: @mswiderski giving your comment the this makes me think that probably this fix is incorrect. I was following the logic based on this

https://github.com/kiegroup/jbpm/blob/7.20.0.Final/jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/ProcessInstanceInfo.java#L194

and copying in the in to the JPAProcessInstanceManager causing the disconnection. The disconnection causes removing the listener at several stages including the one giving problem.

Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageActivationConditionTest.java]: @elguardian In general what I don't see a need is for reading operations we should not make modifications to the underlying entity - process instance info. hat's why I want to understand why we need to do it now. The main reason for it is that it will increase probability for optimistic lock exceptions as even getting process variables will update the version.
With that said, mixing readonly and write operations on process instance should be work as expected - as if I recall correctly readonly does the disconnect as you pointed out but also removes from the internal map of loaded instances and thus should be reloaded when write operation wants to access it. Maybe that would be the area to investigate more?

**Comment by elguardian** [comment]
done @mswiderski created a method for setting the process instance as readonly

File: jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java [hunk 2]
```diff
@@ -174,6 +174,7 @@ public ProcessInstance getProcessInstance(long id, boolean readOnly) {
                 ((ProcessInstanceImpl) processInstance).reconnect();
                 if (readOnly) {
                     internalRemoveProcessInstance(processInstance);
+                    ((ProcessInstanceImpl) processInstance).setReadOnly();
```
Comment (MarianMacik) [inline jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: Doesn't the `setReadOnly` method execute exactly the same method as `internalRemoveProcessInstance(processInstance);`? In `setReadOnly` method there is `...getProcessInstanceManager().internalRemoveProcessInstance(this);` The only addition I see is setting the `process` field to null in the `setReadOnly` method.

Comment (elguardian) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: what is your point ?

Comment (MarianMacik) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: Sorry, maybe the comment was confusing.

I am saying that `setReadOnly()` method calls:

```
public void setReadOnly() {
        ((InternalProcessRuntime) kruntime.getProcessRuntime()).getProcessInstanceManager().internalRemoveProcessInstance(this);
        process = null;
}
```

which is (specifically `internalRemoveProcessInstance(this)`) already called one line above in the code I commented:

```
if (readOnly) {
           internalRemoveProcessInstance(processInstance); // this is duplicated imho
          ((ProcessInstanceImpl) processInstance).setReadOnly();
```

And since the implementation of `internalRemoveProcessInstance()` is:

https://github.com/kiegroup/jbpm/blob/de4094d0023fe11e6188fa86fea85f85c0cddfc3/jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java#L203-L205

Then I think the only real addition made by the new `setReadOnly` method is that it does `process = null;`. So why not only call the new `setReadOnly()` method in that if block?

But maybe there is something more to it which I didn't notice.

Comment (elguardian) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: my bad... I made a mistake with the latest push and the some changes were missing ..

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/subprocess/SubprocessDoublesTest.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        assertProcessInstanceActive(processInstance.getId());
+        assertNodeActive(processInstance.getId(), ksession, "subprocess");
+
+        SubProcessNodeInstance subprocessNode = (SubProcessNodeInstance) processInstance.getNodeInstances().toArray()[0];
+        RuleFlowProcessInstance subprocessInstance = (RuleFlowProcessInstance) ksession
+                                                                                       .getProcessInstance(subprocessNode.getProcessInstanceId());
+
+        // This code simaulates "GET process/instance" rest api call.
+        // Every call of getProcessInstance(id, true) causes one extra-call of firs WorkItem after subprocess.
```
Comment (mswiderski) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/subprocess/SubprocessDoublesTest.java]: I don't understand this comment, what is an extra-call of first work item?

File: jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java [hunk 4]
```diff
@@ -174,6 +174,9 @@ public ProcessInstance getProcessInstance(long id, boolean readOnly) {
                 ((ProcessInstanceImpl) processInstance).reconnect();
                 if (readOnly) {
```
Comment (mswiderski) [inline jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: I am starting to think if we need to make all this changes ... maybe we could simply not reconnect the process instance - not registering any listeners.... as this is just for reading the state of the process instance and should never react to any singals...wdyt?

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/subprocess/SubprocessDoublesTest.java (hunk 3 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/subprocess/SubprocessDoublesTest.java]: it is referred that the the process invokes twices the listener once you get the read only process because it registers twice the same listener

File: jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java (hunk 4 above)
Comment (MarianMacik) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: +1 from me, actually I think we discussed this approach but I wasn't sure if it would break something else or not. Then this would mean to just add else branch for `if (readOnly) {` with the code `((ProcessInstanceImpl) processInstance).reconnect();` I think.

**Comment by elguardian** [comment]
ok to test

File: jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java (hunk 4 above)
Comment (elguardian) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: it does not work.

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by elguardian** [comment]
jenkins retest this

File: jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java (hunk 4 above)
Comment (elguardian) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java]: @MarianMacik it is not enough to reconnect... the listeners can be added anywhere so you have no control where they are added. not all the logic is centralized so it is rather difficult to unravel the whole thing.

**Comment by elguardian** [comment]
jenkins retest this
